metomi/rosie/cli_utils.py

```diff
@@ -0,0 +1,208 @@
+# l, x -> line, cross (i.e. vertex)
+
+# unicode (U_) chars
+U_HL = '─'
+U_VL = '│'
+U_TL = '┌'
+U_TR = '┐'
+U_BL = '└'
+U_BR = '┘'
+U_LX = '├'
+U_RX = '┤'
+U_TX = '┬'
+U_BX = '┴'
+U_XX = '┼'
+
+# ASCII (A_) chars
+A_HL = '-'
+A_VL = '|'
+A_TL = '+'
+A_TR = '+'
+A_BL = '+'
+A_BR = '+'
+A_LX = '+'
+A_RX = '+'
+A_TX = '+'
+A_BX = '+'
+A_XX = '+'
+
+
+def get_terminal_cols(default=80):
+    """Return the width of the terminal in columns.
+
+    If the check fails, this will fallback to the default value.
+    """
+    proc = Popen(['stty', 'size'], stdout=PIPE, stderr=PIPE, text=True)
+    try:
+        return int(proc.communicate()[0].split()[1]) or default
+    except Exception:
+        return default
+
+
+def table(
+    rows: List[List[str]],
+    header: Optional[List[str]] = None,
+    max_width: Optional[int] = None,
+    unicode: bool = True,
+) -> str:
+    """Format text into a table.
+
+    Args:
+        rows: 2D table composed of lists.
+        header: Optional 1D list to use as the table header.
+        max_width: Maximum permitted width for the whole table including
+            borders.
+        unicode: Use unicode characters if True, else fallback to ascii.
+
+    Examples:
+        >>> print(table(['a', 'b', 'c']))
+        ┌───┐
+        │ a │
+        ├───┤
+        │ b │
+        ├───┤
+        │ c │
+        └───┘
+
+        >>> print(table([['a', 'b', 'c']], unicode=False))
+        +---+---+---+
+        | a | b | c |
+        +---+---+---+
+
+        >>> print(table(
+        ...     [['a', 'b', 'c'], ['d', 'e', 'qwertyuiopasdfghjkl']],
+        ...     header=['foo', 'bar', 'baz'],
+        ...     max_width=27,
+        ... ))
+        ┌─────┬─────┬─────────────┐
+        │ foo │ bar │ baz         │
+        ├─────┼─────┼─────────────┤
+        ├─────┼─────┼─────────────┤
+        │ a   │ b   │ c           │
+        ├─────┼─────┼─────────────┤
+        │ d   │ e   │ qwertyuiopa │
+        │     │     │ sdfghjkl    │
+        └─────┴─────┴─────────────┘
+
+    """
+    # determine character set
+    if unicode:
+        hl = U_HL
+        vl = U_VL
+        tl = U_TL
+        tr = U_TR
+        bl = U_BL
+        br = U_BR
+        lx = U_LX
+        rx = U_RX
+        tx = U_TX
+        bx = U_BX
+        xx = U_XX
+    else:
+        hl = A_HL
+        vl = A_VL
+        tl = A_TL
+        tr = A_TR
+        bl = A_BL
+        br = A_BR
+        lx = A_LX
+        rx = A_RX
+        tx = A_TX
+        bx = A_BX
+        xx = A_XX
+
+    # determine column widths
+    widths = [0] * len(rows[0])
+    for table_ in (rows, [header or []]):
+        for row in (table_):
+            for ind, col in enumerate(row):
+                widths[ind] = max(widths[ind], len(col))
+
+    def calc_width():
+        return (sum(widths) + (len(widths) * 3) + 1)
+
+    # resize cols to make them fix the max_width
+    if max_width:
+        overhang = max_width - calc_width()
+        while overhang < 0:
```

Comment:
```python
print(table([["B", "Supercalifragilisticexpialgocious"]], header='hw', max_width=6))
```
dumps you into an infinite loop (`overhang == -1`)

Reply:
Ooh, good spot!

Reply:
Limited to 15 iterations and set a minimum column width of 1.

The table will now exceed the `max_width` in this situation rather than breaking.